File: src/types/transaction.ts
export type TransactionType = "VENTE" | "AVOIR" | "ANNULATION";

export type PaymentMethod =
  | "ESPECES"
  | "CB"
  | "CHEQUE"
  | "TITRE_RESTO"
  | "VIREMENT"
  | "AVOIR"
  | "AUTRE";

export interface Transaction {
  id: string;
  session_id: string;
  sequence_no: number;
  type: TransactionType;
  ref_transaction_id: string | null;
  /** Montants en centimes */
  total_ht: number;
  total_tva: number;
  total_ttc: number;
  discount_ttc: number;
  previous_hash: string | null;
  hash: string;
  receipt_printed: number;
  receipt_email: string | null;
  created_at: string;
}

export interface TransactionLine {
  id: string;
  transaction_id: string;
  line_no: number;
  product_id: string | null;
  product_name: string;
  product_sku: string | null;
  quantity: number;
  /** En centimes */
  unit_price_ttc: number;
  unit_price_ht: number;
  tva_rate_pct: number;
  discount_ttc: number;
  line_total_ht: number;
  line_total_tva: number;
  line_total_ttc: number;
}

export interface Payment {
  id: string;
  transaction_id: string;
  method: PaymentMethod;
  /** En centimes */
  amount: number;
  cash_given: number | null;
  cash_change: number | null;
  reference: string | null;
  created_at: string;
}

export interface TransactionFull {
  transaction: Transaction;
  lines: TransactionLine[];
  payments: Payment[];
}

/** Entrées envoyées par le frontend pour créer une transaction */
export interface CartLineInput {
  product_id: string | null;
  product_name: string;
  product_sku: string | null;
  quantity: number;
  unit_price_ttc: number;
  unit_price_ht: number;
  tva_rate_pct: number;
  discount_ttc: number;
}

export interface PaymentInput {
  method: PaymentMethod;
  amount: number;
  cash_given: number | null;
  reference?: string | null;
}

/**
 * Regroupe les paiements appartenant à une même personne.
 * paymentIndices = positions (dans le tableau PaymentInput[]) des paiements de cette personne.
 * Utilisé uniquement côté frontend pour la gestion des reçus — non envoyé au backend.
 */
export interface PersonGroup {
  label: string;          // "Commande" | "Personne 1" | …
  paymentIndices: number[]; // indices dans le tableau payments de TransactionFull
  total: number;           // centimes
}

File: src-tauri/src/db/models/transaction.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Transaction {
    pub id: String,
    pub session_id: String,
    pub sequence_no: i64,
    pub r#type: String,
    pub ref_transaction_id: Option<String>,
    pub total_ht: i64,
    pub total_tva: i64,
    pub total_ttc: i64,
    pub discount_ttc: i64,
    pub previous_hash: Option<String>,
    pub hash: String,
    pub receipt_printed: i64,
    pub receipt_email: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct TransactionLine {
    pub id: String,
    pub transaction_id: String,
    pub line_no: i64,
    pub product_id: Option<String>,
    pub product_name: String,
    pub product_sku: Option<String>,
    pub quantity: i64,
    pub unit_price_ttc: i64,
    pub unit_price_ht: i64,
    pub tva_rate_pct: i64,
    pub discount_ttc: i64,
    pub line_total_ht: i64,
    pub line_total_tva: i64,
    pub line_total_ttc: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Payment {
    pub id: String,
    pub transaction_id: String,
    pub method: String,
    pub amount: i64,
    pub cash_given: Option<i64>,
    pub cash_change: Option<i64>,
    pub reference: Option<String>,
    pub created_at: String,
}

/// Payload complet renvoyé au frontend pour affichage d'un ticket.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransactionFull {
    pub transaction: Transaction,
    pub lines: Vec<TransactionLine>,
    pub payments: Vec<Payment>,
}

/// Entrée d'une ligne panier envoyée par le frontend.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CartLineInput {
    #[serde(default)]
    pub product_id: Option<String>,
    pub product_name: String,
    #[serde(default)]
    pub product_sku: Option<String>,
    pub quantity: i64,
    pub unit_price_ttc: i64,
    pub unit_price_ht: i64,
    pub tva_rate_pct: i64,
    pub discount_ttc: i64,
}

/// Entrée d'un paiement envoyée par le frontend.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PaymentInput {
    pub method: String,
    pub amount: i64,
    #[serde(default)]
    pub cash_given: Option<i64>,
    #[serde(default)]
    pub reference: Option<String>,
}
